lib/src/back_end/code_writer.dart

```diff
@@ -101,31 +86,16 @@ class CodeWriter {
   /// solution if the line ends up overflowing.
   final List<Piece> _currentUnsolvedPieces = [];
 
-  /// The options for the current innermost piece being formatted.
-  _PieceOptions get _options => _pieceOptions.last;
+  CodeWriter(this._pageWidth, this._solution);
 
-  /// The offset in the formatted code where the selection starts.
+  /// Called after this writer has finished formatting the entire piece tree.
   ///
-  /// This is `null` until the piece containing the selection start is reached
-  /// at which point it gets set. It remains `null` if there is no selection.
-  int? _selectionStart;
-
-  /// The offset in the formatted code where the selection ends.
-  ///
-  /// This is `null` until the piece containing the selection end is reached
-  /// at which point it gets set. It remains `null` if there is no selection.
-  int? _selectionEnd;
-
-  CodeWriter(this._pageWidth, this._pieceStates);
-
-  /// Returns the finished code produced by formatting the tree of pieces and
-  /// the final score.
-  Solution finish() {
+  /// Returns the final formatted text and the next piece that can be expanded
+  /// from this solution, if any.
```

Comment:
[nit] What does "this solution" refer to?

Reply:
The Solution the CodeWriter is bound to.

One way to think of CodeWriter is that it's the API that Piece is given access to in order to format itself for a given solution.

Reworded.

Reply:
This is cleared up by the other thread.